exporter/datadogexporter/internal/metadata/internal/gohai/payload.go

```diff
@@ -0,0 +1,68 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gohai // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/internal/metadata/internal/gohai"
+
+import (
+	"encoding/json"
+)
+
+type gohai struct {
+	CPU        interface{} `json:"cpu"`
+	FileSystem interface{} `json:"filesystem"`
+	Memory     interface{} `json:"memory"`
+	Network    interface{} `json:"network"`
+	Platform   interface{} `json:"platform"`
+}
+
+// Payload handles the JSON unmarshalling of the metadata payload
+// As weird as it sounds, in the v5 payload the value of the "gohai" field
+// is a JSON-formatted string. So this struct contains a MarshaledGohaiPayload
+// which will be marshaled as a JSON-formatted string.
+type Payload struct {
+	Gohai MarshaledGohaiPayload `json:"gohai"`
+}
+
+// MarshaledGohaiPayload contains the marshaled payload
+type MarshaledGohaiPayload struct {
```

Comment:
This structure doesn't hold a marshaled payload (like the name might suggest), it instead implements [`json.Marshaler`](https://pkg.go.dev/encoding/json#Marshaler) and [`json.Unmarshaler`](https://pkg.go.dev/encoding/json#Unmarshaler) on top of a `*gohai` structure. For that reason, let's call it `gohaiSerializer` or `gohaiMarshaler`